File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    # Gives us the new head
    def reverseLL(self, head: Optional[ListNode]) -> Optional[ListNode]:

        curr = head
        prev = None

        while curr:
            next = curr.next

            curr.next = prev
            
            prev = curr
            curr = next

        return prev
        

    def reorderList(self, head: Optional[ListNode]) -> None:
        hare = head
        tortoise = head
        # Jump the hare to the end
        while hare and hare.next and hare.next.next:
            tortoise = tortoise.next
            hare = hare.next.next
        
        sec = tortoise.next
        tortoise.next = None
        secRev = self.reverseLL(sec)
        newTail = head

        while newTail and secRev:
            next = newTail.next
            secNxt = secRev.next

            newTail.next = secRev
            secRev.next = next

            newTail = next
            secRev = secNxt

File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-1.swift
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     var val: Int
 *     var next: ListNode?
 *     init(_ val: Int) {
 *         self.val = val
 *         self.next = nil
 *     }
 * }
 */

class Solution {
    func reverseLL(_ head: ListNode?) -> ListNode? {
        var curr = head
        var prev: ListNode? = nil
        while let c = curr {
            let next = c.next
            c.next = prev
            prev = c
            curr = next
        }
        return prev
    }

    func removeNthFromEnd(_ head: ListNode?, _ n: Int) -> ListNode? {
        // Step 1 - Reverse LL
        let revHead = reverseLL(head)

        // Step 2 - Remove from Index
        let dummyHead = ListNode(-1)
        dummyHead.next = revHead
        var curr = revHead
        var prev = dummyHead

        for _ in 0..<(n-1) {
            guard let c = curr else { break }

            prev = c
            curr = c.next
        }

        if let c = curr, let cn = c.next {
            prev.next = cn
        } else {
            prev.next = nil
        }

        // Step 3 - Reverse back
        return reverseLL(dummyHead.next)
    }
}
